### ISSUE TITLE
Incorrect status hash due to incorrect ordering of same-block transactions

### ISSUE THREAD

**Issue by ghost** [romanz/electrs#297]
Hi,

currently there is an bug in the implementation of blockchain.scripthash.subscribe.
Normally what i would expect if i subscribe to an script hash / address, i should get only notified on changes.
However i am receiving the old status permanently.

> https://www.blockchain.com/btc/address/374ZnJhojVvg72MpPsmVx6Yo2zWoGR8s49
> https://electrumx.readthedocs.io/en/latest/protocol-methods.html
> The client will receive a notification when the status of the script hash changes.

https://www.blockchain.com/btc/address/374ZnJhojVvg72MpPsmVx6Yo2zWoGR8s49

Build on Ubuntu 18.04, version 0.8.5.

electrs log:
```
2020-09-14T09:41:28.279+00:00 - TRACE - RPC Request("{\"json-rpc\":\"2.0\",\"id\":6346890332372894945,\"method\":\"blockchain.scripthash.subscribe\",\"params\":[\"f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3\"]}\r\n")
2020-09-14T09:41:30.667+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:41:35.707+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:41:40.750+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:41:45.790+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:41:50.827+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:41:55.855+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:42:00.880+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:42:05.913+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:42:10.948+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:42:15.981+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:42:21.025+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:42:26.058+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:42:31.103+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:42:36.133+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:42:41.173+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:42:46.215+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:42:51.269+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:42:56.312+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:43:01.352+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:43:06.394+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:43:11.436+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:43:16.478+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:43:21.510+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:43:26.553+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:43:31.600+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:43:36.641+00:00 - TRACE - RPC PeriodicUpdate
2020-09-14T09:43:41.678+00:00 - TRACE - RPC PeriodicUpdate
...
```

client log:
```
andreas@aw-ubuntu:~/Apps$ telnet 127.0.0.1 50001
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
{"json-rpc":"2.0","id":6346890332372894945,"method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3"]}
{"id":6346890332372894945,"jsonrpc":"2.0","result":"c310602da82d3cd6f0ccc680aa39568354defbe91e21717c29a10fb71c976b70"}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","a3707437d6ae71046282a6e9f509c88022527abc68e7582afa55497e0e5e9d02"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","c310602da82d3cd6f0ccc680aa39568354defbe91e21717c29a10fb71c976b70"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","a3707437d6ae71046282a6e9f509c88022527abc68e7582afa55497e0e5e9d02"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","c310602da82d3cd6f0ccc680aa39568354defbe91e21717c29a10fb71c976b70"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","a3707437d6ae71046282a6e9f509c88022527abc68e7582afa55497e0e5e9d02"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","c310602da82d3cd6f0ccc680aa39568354defbe91e21717c29a10fb71c976b70"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","a3707437d6ae71046282a6e9f509c88022527abc68e7582afa55497e0e5e9d02"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","c310602da82d3cd6f0ccc680aa39568354defbe91e21717c29a10fb71c976b70"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","a3707437d6ae71046282a6e9f509c88022527abc68e7582afa55497e0e5e9d02"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","c310602da82d3cd6f0ccc680aa39568354defbe91e21717c29a10fb71c976b70"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","a3707437d6ae71046282a6e9f509c88022527abc68e7582afa55497e0e5e9d02"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","c310602da82d3cd6f0ccc680aa39568354defbe91e21717c29a10fb71c976b70"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","a3707437d6ae71046282a6e9f509c88022527abc68e7582afa55497e0e5e9d02"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","c310602da82d3cd6f0ccc680aa39568354defbe91e21717c29a10fb71c976b70"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","a3707437d6ae71046282a6e9f509c88022527abc68e7582afa55497e0e5e9d02"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","c310602da82d3cd6f0ccc680aa39568354defbe91e21717c29a10fb71c976b70"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","a3707437d6ae71046282a6e9f509c88022527abc68e7582afa55497e0e5e9d02"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","c310602da82d3cd6f0ccc680aa39568354defbe91e21717c29a10fb71c976b70"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","a3707437d6ae71046282a6e9f509c88022527abc68e7582afa55497e0e5e9d02"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","c310602da82d3cd6f0ccc680aa39568354defbe91e21717c29a10fb71c976b70"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","a3707437d6ae71046282a6e9f509c88022527abc68e7582afa55497e0e5e9d02"]}
{"jsonrpc":"2.0","method":"blockchain.scripthash.subscribe","params":["f8822faf36af873f8ab49a837f64c213b3a884289995644286a56dee112ac1d3","c310602da82d3cd6f0ccc680aa39568354defbe91e21717c29a10fb71c976b70"]}
```
Kind regards,
lamp-aw

**Comment by craigraw** [comment]
Seeing the same here (Electrs 0.8.5). In my case Electrs picks two subscribed script hashes (/0/0 and /0/8, no idea why) and keeps sending updates for them. The script hash passed in bounces between two statuses - the current status and a previous status. From my logging:

```
14:14:09,581 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd now 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198
14:14:09,581 INFO Status updated for script hash 5f630b635be6d288fdf987cbb9e82b9751f3fa13de0c56f32fd74c8c13cf88b7, was 2f6d31e41cd2b853f489ea34b2261bcc9463ec5c078a4d3ad480f6c99102036f now 9164b544a1de2f3c97a4ddfc0a929641e0c270e7a2030c5b656a5e6a31a314d7
14:14:19,927 INFO Status updated for script hash 5f630b635be6d288fdf987cbb9e82b9751f3fa13de0c56f32fd74c8c13cf88b7, was 9164b544a1de2f3c97a4ddfc0a929641e0c270e7a2030c5b656a5e6a31a314d7 now 2f6d31e41cd2b853f489ea34b2261bcc9463ec5c078a4d3ad480f6c99102036f
14:14:25,050 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198 now 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd
14:14:30,109 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd now 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198
14:14:45,644 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198 now 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd
14:14:45,645 INFO Status updated for script hash 5f630b635be6d288fdf987cbb9e82b9751f3fa13de0c56f32fd74c8c13cf88b7, was 2f6d31e41cd2b853f489ea34b2261bcc9463ec5c078a4d3ad480f6c99102036f now 9164b544a1de2f3c97a4ddfc0a929641e0c270e7a2030c5b656a5e6a31a314d7
14:14:50,579 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd now 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198
14:14:55,772 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198 now 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd
14:15:06,217 INFO Status updated for script hash 5f630b635be6d288fdf987cbb9e82b9751f3fa13de0c56f32fd74c8c13cf88b7, was 9164b544a1de2f3c97a4ddfc0a929641e0c270e7a2030c5b656a5e6a31a314d7 now 2f6d31e41cd2b853f489ea34b2261bcc9463ec5c078a4d3ad480f6c99102036f
14:15:11,353 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd now 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198
14:15:16,506 INFO Status updated for script hash 5f630b635be6d288fdf987cbb9e82b9751f3fa13de0c56f32fd74c8c13cf88b7, was 2f6d31e41cd2b853f489ea34b2261bcc9463ec5c078a4d3ad480f6c99102036f now 9164b544a1de2f3c97a4ddfc0a929641e0c270e7a2030c5b656a5e6a31a314d7
14:15:21,782 INFO Status updated for script hash 5f630b635be6d288fdf987cbb9e82b9751f3fa13de0c56f32fd74c8c13cf88b7, was 9164b544a1de2f3c97a4ddfc0a929641e0c270e7a2030c5b656a5e6a31a314d7 now 2f6d31e41cd2b853f489ea34b2261bcc9463ec5c078a4d3ad480f6c99102036f
14:15:32,228 INFO Status updated for script hash 5f630b635be6d288fdf987cbb9e82b9751f3fa13de0c56f32fd74c8c13cf88b7, was 2f6d31e41cd2b853f489ea34b2261bcc9463ec5c078a4d3ad480f6c99102036f now 9164b544a1de2f3c97a4ddfc0a929641e0c270e7a2030c5b656a5e6a31a314d7
14:15:37,388 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198 now 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd
14:15:42,567 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd now 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198
14:15:42,568 INFO Status updated for script hash 5f630b635be6d288fdf987cbb9e82b9751f3fa13de0c56f32fd74c8c13cf88b7, was 9164b544a1de2f3c97a4ddfc0a929641e0c270e7a2030c5b656a5e6a31a314d7 now 2f6d31e41cd2b853f489ea34b2261bcc9463ec5c078a4d3ad480f6c99102036f
14:15:58,239 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198 now 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd
14:16:03,394 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd now 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198
14:16:13,737 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198 now 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd
14:16:18,948 INFO Status updated for script hash 5f630b635be6d288fdf987cbb9e82b9751f3fa13de0c56f32fd74c8c13cf88b7, was 2f6d31e41cd2b853f489ea34b2261bcc9463ec5c078a4d3ad480f6c99102036f now 9164b544a1de2f3c97a4ddfc0a929641e0c270e7a2030c5b656a5e6a31a314d7
14:16:24,063 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd now 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198
14:16:29,267 INFO Status updated for script hash 3547064c78e7fb97abd824eef9a5133ff2ebded1b34c35a9bf0c10f359f9a8a3, was 6800dee896d68ed3ed32386e1f35e3a22d4b6570268f4973b814f5a77df1f198 now 31a2bce884a91755140aec0f301b7c47d82266f29b34e07997bf6e2eec06cffd
14:16:29,268 INFO Status updated for script hash 5f630b635be6d288fdf987cbb9e82b9751f3fa13de0c56f32fd74c8c13cf88b7, was 9164b544a1de2f3c97a4ddfc0a929641e0c270e7a2030c5b656a5e6a31a314d7 now 2f6d31e41cd2b853f489ea34b2261bcc9463ec5c078a4d3ad480f6c99102036f
```

I observe the same behaviour in @lamp-aw's logging.

**Comment by craigraw** [comment]
Confirmed the same behaviour occurs with this wallet in Electrum 4.0.3.

**Comment by romanz** [comment]
Thanks for reporting this issue!
Can be reproduced on current `master`:
```
$ contrib/addr.py 374ZnJhojVvg72MpPsmVx6Yo2zWoGR8s49
[{'height': 641371, 'tx_hash': '34a44698999e49d3c436616d212d06a61a40c478ee1d8365da80e0a8bf269358'}, {'height': 641371, 'tx_hash': '5a7c9d25ba2db4324cd58f9502149f86b33a4ac7094a3ddfb2dedd9825c469eb'}]
$ contrib/addr.py 374ZnJhojVvg72MpPsmVx6Yo2zWoGR8s49
[{'height': 641371, 'tx_hash': '5a7c9d25ba2db4324cd58f9502149f86b33a4ac7094a3ddfb2dedd9825c469eb'}, {'height': 641371, 'tx_hash': '34a44698999e49d3c436616d212d06a61a40c478ee1d8365da80e0a8bf269358'}]
```
The current code doesn't sort the transactions' history correctly:
https://electrumx-spesmilo.readthedocs.io/en/latest/protocol-basics.html#status
```
order confirmed transactions to the script hash by increasing height 
(and position in the block if there are more than one in a block)
```

**Comment by romanz** [comment]
Should be resolved in [`p2p` branch](https://github.com/romanz/electrs/commits/p2p/):
https://github.com/romanz/electrs/blob/5d3457b764778cffebe991ce5aab812b85a1e30e/src/status.rs#L278

**Comment by dmp1ce** [comment]
I had a problem with a legacy testnet wallet which had some mined coins. From `master` I would not sync fully. With `p2p` it synced immediately! When is `p2p` going to be merged to master? It seems to work great?

BTW, I used the Dockerfile from master on the `p2p` branch in order to get `p2p` working on my electrs node.

**Comment by romanz** [comment]
Cool, glad to hear!
I hope to merge it soon - but first I need to add more tests and features - e.g. https://github.com/romanz/electrs/pull/390#issuecomment-822515354.

**Comment by romanz** [comment]
Resolved in latest 0.9.0 release :)
